rfc/rfc-0000-allow-resources-to-report-status.md

```diff
@@ -0,0 +1,298 @@
+- RFC Pull Request: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+The RFC proposes modifying the existing templates to allow template authors to indicate how Cartographer can read the state of the resource so that it can be reflected on the owner status.
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+  - Currently, a blueprint has no knowledge of the current status of the stamped resources. This RFC proposes how resources can report (and categorize) this information strictly for informational purposes.
+- What use cases does it support?
+  - This feature will surface information about the health of the underlying resources directly on the owner.
+  - This will also enable integrations that may be looking to create a UI around Cartographer, allowing the UI to display the health of the blueprint without having to make an API call to k8s to retrieve the status of every resource, and without having to know how to define health for every stamped object type.
+- What is the expected outcome?
+  - Users will have more insight into the progression of the blueprint and the health of each underlying resource.
+  - This is augmenting information exposed via resource tracing, and is not a solution for artifact tracing.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes adding a field to the spec of all templates, `healthRule`. Template authors use the field to define the conditions or fields on a stamped resource that determine whether the resource is healthy.
+
+Cartographer will attempt to read the specified conditions and/or fields and reflect this state on the owner status. If Cartographer cannot determine health, it will show the status as unknown.
+
+Single condition example (likely the most common use case):
+```yaml
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterSourceTemplate
+metadata:
+  name: source
+spec:
+  urlPath: .status.artifact.url
+  revisionPath: .status.artifact.revision
+
+  healthRule:
+    singleConditionType: Ready
+
+  template:
+    apiVersion: source.toolkit.fluxcd.io/v1beta1
+    kind: GitRepository
+
+---
+apiVersion: carto.run/v1alpha1
+kind: Workload
+status:
+  resources:
+    - name: source-provider
+      conditions:
+        - type: Healthy
+          status: 'True'
+          message: 'Fetched revision: main/...'
+          reason: ReadyCondition
+          lastTransitionTime: "2022-03-23T13:08:22Z"
+    - name: source-provider-2
+      conditions:
+        - type: Healthy
+          status: 'Unknown'
+```
+
+Multi match with both conditions and fields, plus `messagePath` in matchFields to allow for surfacing the most useful messages:
+```yaml
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterTemplate
+metadata:
+  name: deploy
+spec:
+  healthRule:
+    multiMatch:
+      healthy:
+        matchConditions:
+          - type: ReconcileSucceeded
+            status: 'True'
```

Comment:
it's interesting that a resource like `kapp-controller/App` gives us conditions (which would be useful for either `singleCondition` or `multiMatch` w/ `matchConditions`, but as there's no message/reason under a `ReconcileSucceeed: True` scenario

```yaml
status:
  friendlyDescription: Reconcile succeeded

  conditions:
  - status: "True"              # no message/reason fields
    type: ReconcileSucceeded

  deploy:
    exitCode: 0
    finished: true
    startedAt: "2022-04-12T18:47:21Z"
    stdout: |-
      Target cluster 'https://10.96.0.1:443'
      Changes
      Namespace  Name  Kind     Conds.  Age  Op      Op st.  Wait to    Rs  Ri
      default    app   Service  -       -    create  -       reconcile  -   -
      Op:      1 create, 0 delete, 0 update, 0 noop, 0 exists
      Wait to: 1 reconcile, 0 delete, 0 noop
      6:47:21PM: ---- applying 1 changes [0/1 done] ----
      Pod watching error: Creating Pod watcher: unknown (get pods)
      6:47:21PM: create service/app (serving.knative.dev/v1) namespace: default
      6:47:21PM: ---- waiting on 1 changes [0/1 done] ----
      6:47:21PM: ok: reconcile service/app (serving.knative.dev/v1) namespace: default
      6:47:21PM: ---- applying complete [1/1 done] ----
      6:47:21PM: ---- waiting complete [1/1 done] ----
      Succeeded
...
```

we end up actually being somewhat force to use `multiMatch` with `matchFields` for the `healthy` case too for a complete representation that'd include the `status.friendlyDescription` I believe

Reply:
`App` will already require the `multiMatch` as it uses independent `Success` and `Failure` conditions (which is a nit I'll pick another day). Even if `App` used a single status condition, I think it's fair to drop down from the `singleCondition` support to pick up the custom message.

`singleCondition` is syntax sugar rather than defining a new capability.